WorkflowSwiftUI/Sources/Store.swift

```diff
@@ -0,0 +1,453 @@
+        }
+    }
+}
+
+// MARK: - Scoping
+
+public extension Store {
+    /// Holds a cached child store for a nested ObservableModel on this store's model.
+    internal struct ChildStore {
+        var store: Any
+        var setModel: (Model) -> Void
+        var isInvalid: (Model) -> Bool
+
+        private var _invalidate: () -> Void
+
+        init(
+            store: Store<some ObservableModel>,
+            setModel: @escaping (Model) -> Void,
+            isInvalid: @escaping (Model) -> Bool
+        ) {
+            self.store = store
+            self.setModel = setModel
+            self.isInvalid = isInvalid
+
+            self._invalidate = {
+                store.invalidate()
+            }
+        }
+
+        func invalidate() {
+            _invalidate()
+        }
+    }
+
+    /// Represents an access to the "wrapper" of a nested child store, such as an Optional or
+    /// collection type.
+    ///
+    /// Each nested model's scope will track its own mutations, but we use this to track mutations
+    /// to the wrapper itself, such as changes to a collection size.
+    internal struct ChildModelAccess {
+        var willSet: (Store<Model>) -> Void
+        var didSet: (Store<Model>) -> Void
+        var changed: (Model, Model) -> Bool
+        var isInvalid: (Model) -> Bool
+
+        init(
+            keyPath: KeyPath<Model, some Any>,
+            changed: @escaping (Model, Model) -> Bool,
+            isInvalid: @escaping (Model) -> Bool
+        ) {
+            self.willSet = { store in
+                store._$observationRegistrar.willSet(store, keyPath: (\Store.model).appending(path: keyPath))
+            }
+            self.didSet = { store in
+                store._$observationRegistrar.didSet(store, keyPath: (\Store.model).appending(path: keyPath))
+            }
+            self.changed = changed
+            self.isInvalid = isInvalid
+        }
+    }
+
+    /// Track access to a child store wrapper.
+    internal func access(
+        keyPath key: KeyPath<Model, some Any>,
+        changed: @escaping (Model, Model) -> Bool,
+        isInvalid: @escaping (Model) -> Bool = { _ in false }
+    ) {
+        _$observationRegistrar.access(self, keyPath: (\Store.model).appending(path: key))
+        if childModelAccesses[key] == nil {
+            childModelAccesses[key] = ChildModelAccess(
+                keyPath: key,
+                changed: changed,
+                isInvalid: isInvalid
+            )
+        }
+    }
+
+    internal func scope<ChildModel>(
+        key: AnyHashable,
+        getModel: @escaping (Model) -> ChildModel,
+        isInvalid: @escaping (Model) -> Bool
+    ) -> Store<ChildModel> {
+        if let childStore = childStores[key]?.store as? Store<ChildModel> {
+            return childStore
+        }
+
+        let childModel = getModel(model)
+        let childStore = Store<ChildModel>(childModel)
+
+        childStores[key] = ChildStore(
```

Comment:
Should this be asserting there isn't already a value stored by this key?

Reply:
At the top of the function we return the cached store if there's already a store for this key.

Reply:
Oh, doh